Resolves Issue #227 - XML files need to have better error handling

spaulaus opened the pull request:
This pull request resolves Issue #227. This is a nearly complete overhaul to the handling of XML files. There are new classes that are dedicated to parsing specific nodes of the configuration file. For example, the Map node is now parsed once from a single file instead of three times from two files. The Globals class has been gutted of any kind of XML parsing, it also has the constants and functions removed from the header.

I have now added a XmlInterface and XmlParser, to the resources. These facilitate other programs to use XML files if they choose.

spaulaus committed:
Adding header that contains functions related to string manipulation

These will replace the string namespace in the utkscan Globals header. This will allow more classes to take advantage of these useful functions.

spaulaus committed:
Removing string namespace from utkscan::Globals

This namespace has now been removed. Several classes have been updated to take advantage of the new functions.

spaulaus committed:
Most of the constants defined in the Globals class were unused

They were legacy values from when the unpacking was still done by PixieStd.cpp. The one constant that was useful, maximum number of channels on a pixie card, was moved to Resources::Constants and put under the Pixie16 namespace. The remaining references have been replaced by calls to the appropriate numeric_limits methods since they were there to replace UINT_MAX and the other associated values.

spaulaus committed:
Removing the unnecessary NAN definition. 

Nothing in the software used this definition. It is clutter.

spaulaus committed:
XmlInterface compiles without issues

This new class will handle the program's interface with the XML documents. This removes the responsibility from the Globals class in Utkscan, and provides more programs access to this. 

The class is currently a singleton since I do not know of any programs that need to access more than one XML file at a time in the current software base.

spaulaus committed:
Globals has been updated to use the XmlInterface class

This removes the responsibility of opening XML files from the Globals class.

spaulaus committed:
Adding a header that contains various constants.

These constants are things that users will likely never need to modify. For example, the length of a VANDLE bar in cm or the number of channels in a Pixie-16 module.

spaulaus committed:
Adding a function to convert seconds to different time units.

This currnetly only converts to some of the metric prefixes. It could be expanded later to consider different values.

spaulaus committed:
Updating messages in the XmlInterface 

This message is now clearer as to what the interface has actually done.

spaulaus committed:
Removing Xml Parsing from Globals into XmlParser class

This parser handles parsing the XML based configuration file. All of the parsing that the Globals class used to do is now contained here. Globals checks that the necessary nodes are not empty and then sends them to the XmlParser for processing. 

We have added getters and setters to Globals instead of accessing them through the previous member functions. This brings us more in line with standard programming practices. 

XmlParser has had the parsing methods updated from the previous usage in Globals. These methods now include more error checking, and will throw if they are missing information that is critial to the program. They will also output more infromation to the user about what values are actually being used.

spaulaus committed:
Updating analyzers to use the new Globals methods.

spaulaus committed:
Updating utkscan Core to use the new Globals methods

spaulaus committed:
Updating EventProcessors to use the new Globals Methods.

spaulaus committed:
Updating pugixml source code

spaulaus committed:
Abstracting the XmlParser and moving to Resources

This class actually has some use to other programs and classes. I have abstracted the class and moved it to Resources so that it can be used else where. This allows us to define error checking methods for all of the XML based readers.

spaulaus committed:
Making the error checking methods protected instead of private

spaulaus committed:
Adding GlobalsXmlParser

The GlobalsXmlParser handles parsing nodes in the utkscan configuration that are needed for the Globals' class. This will also let us create more of these parsers for nodes that are not handled by the Globals class. This will help keep the length of any one file from getting too long.

spaulaus committed:
Updaing the XmlInterface error message to contian pugixml info

spaulaus committed:
The MapNodeXmlParser has taken over from DetectorLibrary

The messaging has been updated, and some of the code has been rearranged to make it more efficient. There seems to be no issues with parsing the node.

spaulaus committed:
MapNodeXmlParser handles everything in the map node. 

I the parsing here is much improved over the previous iteration. The Walk Correction and the Calibrations are now handled by the same subroutine. Error handing and messaging is also much better. This passes off the constructed Calibrator and WalkCorrector objects to be owned by the DetectorLibrary, which should persist until the end of the program life. 

I updated DetectorLibrary to take thes WalkCorrector and Calibrator objects. 

DetectorDriver obtains pointers to the Calibrator and WalkCorrector objects from DetectorLibrary. This should cut down on the overhead necessary for these classes. The ReadCalXml and ReadWalkXml methods have been removed.

spaulaus committed:
Processors are now parsed and loaded via the DetectorDriverXmlParser

This class handles parsing the DetectorDriver node from the configuration file. It does some error handling and then provides the vectors with the pointers back to DetectorDriver. DetectorDriver is no longer in the business of parsing the XML file.

spaulaus committed:
Notebook is now updated to use the XmlInterface

The Notebook class doesn't really do much with the XML file. i do not think that it warrants a dedicated class since we'll just end up adding more lines of code in this case.

spaulaus committed:
XML parsing removed from TreeCorrelator and into TreeCorrelatorXmlParser

This parser handles parsing the TreeCorrelator node and adding the information to the TreeCorrelator. This  may require some cleanup later but it is working as expected.

spaulaus committed:
Removing the last remains of XML from the Globals Class.

The globals class is not completely independent of any XML parsing.

spaulaus committed:
Now loading the configuration in UtkScanInterface::Initalize

This allows us to perform error handling before we get too far into processing. This also makes the workflow clearer since things like TreeCorrelator are not instanced in the DetectorLibrary.

spaulaus committed:
Fixing error message so that it outputs the proper node name.

spaulaus committed:
Updating TraceAnalyzers so that it is not initalized before everything

The static initalization of the Plots class in the TraceAnalyzer was causing a Static Initialization Order Fiasco and bypassing the error handling in the XML parsing. This fixes the last remaining of these issues that were originally identificed in Issue #13.

spaulaus committed:
Fixing the TraceExtractor so that it plots traces properly.

spaulaus committed:
Changing how the output file is handled. Provides more flexibility.

We previously didn't handle the input provided by -o cleanly. We simply had the entire output file path and didn't differentiate between the file and the path. The path is necessary to ensure that all output flows to the same place. The raw file name is needed so that additional output files can be given the same name. 

The output path no longer needs to be read from the utkscan XML configuration file.

spaulaus committed:
Updating processor to use the appropriate output methods.

spaulaus committed:
Fixing bug in GlobalsXmlParser where fitting parameters weren't read.

mvostinar commented on Analysis/Utkscan/core/source/TimingCalibrator.cpp:
```diff
@@ -32,7 +32,7 @@ TimingCalibrator* TimingCalibrator::get() {
 }
 
 void TimingCalibrator::ReadTimingCalXml() {
```
This should follow the same programing as the Notebook since it is reopening the same .xml file again.

spaulaus replied:
I updated this to follow the same programming as in Notebook. This class needs to be updated to handle the geometric correction a little differently, but this will bring it in line with the scope of this pull request.

mvostinar reviewed (dismissed):
There is an issue which needs to be addressed, see the comment in the file. Otherwise the code seems fine.

spaulaus committed:
Updating issues found during review.

spaulaus committed:
Reorganising the order that include directories happen.

There was an issue found during functional reivew where the Resources were  unable to find the Third Party headers since they were included after the Resources were compiled.

spaulaus committed:
Adding missing header file to the XmlParser. 

invalid_argument was throwing a compilation error.

mvostinar reviewed (approved):
Ubuntu version 14.04.1 -#64
Kernel number 3.19.0-58-generic
Cmake version 3.7.1
gcc version 4.8.4

It passed all the testing that was reported in the issue #227. The program was correctly issuing the error messages when made on purpose. It was behaving in the expected way for all the tests. One additional side possible issue has been find which can be addressed in the future.
